### DISCUSSION TITLE
Info on Protocols

### DISCUSSION THREAD

**Discussion by BootROMError8046** [keirf/flashfloppy#556] [General]
Hi there, first and foremost thank you for your hard work and for your amazing firmware!! I don't know if this is the right place to ask so sorry in advance. I'm taking an electrical engineering course in college and I've chosen to make an open implementation of the gotek so I have to write a firmware for it from scratch.

I you don't mind I have a few questions:

1. I couldn't find any evaluation boards based on the STM32F105RBT6, but I do have a necleo64 which is based on the STM32F103RBT6, so development on the F103 should be compatible with the F105 minus the USB OTG feature, right?
2. Do you have good sources of information about the floppy protocol and how it communicates with the floppy controller on a PC (My main focus is to get my firmware to communicate with a PC as of now, amigas, rolands etc. will have to come later, and I intend to work only with 1.44MB images now).
3. Do you have some tips and tricks on how to develop the firmware? Unfortunately I have to learn as I go and I'm not experienced enough with C so I'm struggling to read your code as of now. 
4. Can you briefly explain how the USB side works in software? Like how the floppy image manager formats the USB stick and makes folders for each image and stuff?

Thank you so much and have a nice day! :)

**Comment by keirf** [comment]
1. Yes though the F103 also has less RAM. Perhaps not a problem to get you going though. Lack of OTG does mean you can't talk to a USB stick of course.
2. The information is rather spread out across datasheets and the like. You need to learn how the data gets framed into sectors, with an IDAM (header) and DAM (data area) each, and then the MFM encoding. Working this out is probably the trickiest bit given the lack of cohesive docs.
3. Perhaps you could start with STMcube? I think it might give you a framework you can then add your own C code into?
4. FlashFloppy works with image files in a normal FAT32 filesystem. So there is no floppy image manager.

**Comment by ejona86** [comment]
For 2, I'd recommend starting with http://www.moria.de/~michael/floppy/ to provide a view of the many components. Then "choose a drive" and look at its manual. They pretty similar. Here's an arbitrary one https://archive.org/details/bitsavers_shugartSA880_2086224

**Comment by keirf** [comment]
Oh yes I forgot about the floppy user guide. That's pretty good!

**Comment by BootROMError8046** [comment]
Hi everyone, sorry for the late reply, thank you both keirf and ejona86! Your help is greatly appreciated! For the ram issue with the F103: Do you think it could be an issue? And what about the floppy manager software? Can flashfloppy mount images I drag'n'dropped to the usb's root directory? Despite owning a gotek for years I actually never had the change to change the firmware on it. Also thank you ejona86 for the documentation! I shall give it a read soon.
